File: topic/construct_binary_tree_from_preorder_and_inorder_traversal/construct_binary_tree_from_preorder_and_inorder_traversal.go
package construct_binary_tree_from_preorder_and_inorder_traversal

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func BuildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) == 0 {
		return nil
	}

	idx := 0
	for i, val := range inorder {
		if val == preorder[0] {
			idx = i
			break
		}
	}

	root := &TreeNode{
		Val:   preorder[0],
		Left:  BuildTree(preorder[1:idx+1], inorder[:idx]),
		Right: BuildTree(preorder[idx+1:], inorder[idx+1:]),
	}

	return root
}

File: topic/construct_binary_tree_from_preorder_and_inorder_traversal/cpp/construct_binary_tree_from_preorder_and_inorder_traversal.cpp
#include <vector>
using namespace std;

struct TreeNode
{
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode() : val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
	TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
	TreeNode *build(vector<int> &preorder, vector<int> &inorder, int stop, int &prePos, int &inPos)
	{
		if (prePos >= preorder.size())
			return nullptr;

		if (inorder[inPos] == stop)
		{
			inPos++;
			return nullptr;
		}

		TreeNode *node = new TreeNode(preorder[prePos++]);
		node->left = build(preorder, inorder, node->val, prePos, inPos);
		node->right = build(preorder, inorder, stop, prePos, inPos);

		return node;
	}

	TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder)
	{
		int inPos = 0, prePos = 0;
		return build(preorder, inorder, INT_MIN, prePos, inPos);
	}
};